lms/extractors/base.py

```diff
@@ -38,37 +50,38 @@ def _split_header(cls, code: CodeFile) -> Tuple[str, str]:
 
         clean_text = code.strip('#' + string.whitespace)
         first_line_end = clean_text.find('\n')
-        first_line = clean_text[:first_line_end].strip()
+        if first_line_end == -1:
+            first_line_end = len(clean_text)
+        first_line = clean_text[:first_line_end].strip().replace('_', ' ')
         code_lines = clean_text[first_line_end:].strip()
 
         logger.debug(f'Upload title: {first_line}')
         return first_line, code_lines
 
     @classmethod
-    def _clean(cls, code: Union[Sequence, str]) -> Tuple[str, str]:
+    def _clean(cls, code: Union[Sequence, str]) -> Tuple[int, str]:
         first_line, code_text = cls._split_header(code)
         upload_title = cls.UPLOAD_TITLE.fullmatch(first_line)
         if upload_title:
-            return upload_title.group(1), code_text
+            exercise_id = int(upload_title.group(1))
+            return exercise_id, code_text
 
         logger.debug(f'Unmatched title: {first_line}')
-        return '', ''
+        return 0, ''
 
-    @abstractmethod
     def can_extract(self) -> bool:
-        pass
+        raise NotImplementedError()
 
-    @classmethod
-    @abstractmethod
-    def get_exercise(cls, to_extract: Any) -> Tuple[str, str]:
-        pass
+    def get_exercise(self, to_extract: Any) -> Tuple[int, List[File]]:
+        raise NotImplementedError()
 
-    @abstractmethod
-    def get_exercises(self):
-        pass
+    def get_exercises(self) -> Iterator[Tuple[int, List[File]]]:
+        raise NotImplementedError()
 
-    def __iter__(self) -> Iterator[Tuple[str, str]]:
+    def __iter__(self) -> Iterator[Tuple[int, List[File]]]:
         for cls in self.__class__.__subclasses__():
+            logger.debug(f'Trying extractor: {cls.__name__}')
             extractor = cls(to_extract=self.to_extract)
             if extractor.can_extract():
```

Comment:
* what happens when multiple extractors can extract? you will continue to get `(solution_id, files)` on the same file and you may encounter a weird behavior where `for (solution_id, files) in extractor:` will give you duplicates.

* another thing, i think it will be best if we could have a log entry saying "i couldn't extract this: {filename}"

my suggestion is to do something like:
```python
for cls ....:
    if extractor.can_extract():
        for ...
            yield ...
       return
    logger.warning(f"couldn't find extractor for {self.to_extract}")
```